lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
+        fprintf(out, " %0.4f", self->root_distribution[j]);
+    }
+    fprintf(out, "\n\t------------------------------------\t\n");
+    for (j = 0; j < self->alphabet_length; j++) {
+        mutation_row = self->mutation_matrix + j * self->alphabet_length;
+        for (k = 0; k < self->alphabet_length; k++) {
+            fprintf(out, " %0.4f", mutation_row[k]);
+        }
+        fprintf(out, "\n");
+    }
+    fprintf(out, "\n");
+}
+
+static int MSP_WARN_UNUSED
+mutation_model_check_validity(mutation_model_t *self)
+{
+    int ret = 0;
+    int j, k;
+    double prob, min_prob;
+    double *mutation_row;
+
+    /* Check probabilities sum to one */
+    prob = 0;
+    min_prob = 0;
+    for (j = 0; j < self->alphabet_length; j++) {
+        prob += self->root_distribution[j];
+        min_prob = GSL_MIN(min_prob, self->root_distribution[j]);
+    }
+    if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+        ret = MSP_ERR_BAD_ROOT_PROBABILITIES;
+        goto out;
+    }
+    for (j = 0; j < self->alphabet_length; j++) {
+        prob = 0;
+        min_prob = 0;
+        mutation_row = self->mutation_matrix + j * self->alphabet_length;
+        for (k = 0; k < self->alphabet_length; k++) {
+            prob += mutation_row[k];
+            min_prob = GSL_MIN(min_prob, mutation_row[k]);
+        }
+        if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+            ret = MSP_ERR_BAD_MUTATION_MATRIX;
+            goto out;
+        }
+    }
+
+out:
+    return ret;
+}
+
+int MSP_WARN_UNUSED
+mutation_model_alloc(mutation_model_t *self, int alphabet_length,
+        const char *alphabet, double *root_distribution, double *mutation_matrix)
+{
+    int ret = 0;
+    int j, k, jk;
+    char *buff = NULL;
+
+    memset(self, 0, sizeof(mutation_model_t));
+    if (alphabet_length < 2) {
+        ret = MSP_ERR_INSUFFICIENT_ALLELES;
+        goto out;
+    }
+
+    buff = malloc((size_t) alphabet_length * sizeof(*self->alphabet));
+    self->root_distribution = malloc((size_t) alphabet_length * sizeof(*self->root_distribution));
+    self->mutation_matrix = malloc((size_t) (alphabet_length * alphabet_length)
+                                            * sizeof(*self->mutation_matrix));
+    if (self->root_distribution == NULL
+            || self->mutation_matrix == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    self->alphabet_length = alphabet_length;
+    for (j = 0; j < alphabet_length; j++) {
+        buff[j] = alphabet[j];
+        self->root_distribution[j] = root_distribution[j];
+        for (k = 0; k < alphabet_length; k++) {
+            jk = j * alphabet_length + k;
+            self->root_distribution[jk] = mutation_matrix[jk];
+        }
+    }
+    self->alphabet = buff;
+    ret = mutation_model_check_validity(self);
+out:
+    return ret;
+}
+
+int
+mutation_model_free(mutation_model_t *self)
+{
+    if (self->alphabet != NULL) {
```

Comment:
This is what msp_safe_free does. Why are you freeing the memory pointed to by self->alphabet?

Reply:
well, since I malloc'ed it? It wouldn't let me `msp_safe_free( )` it because it's `const`. But if I do
```
self->alphabet = alphabet;
```
then maybe I don't need to free it... I'll try... ?